app/models/life_cycle.rb

```diff
@@ -0,0 +1,47 @@
+#-- copyright
+# OpenProject is an open source project management software.
+# Copyright (C) the OpenProject GmbH
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+class LifeCycle < ApplicationRecord
+  has_many :project_life_cycles, class_name: "Project::LifeCycle", dependent: :destroy
+  has_many :projects, through: :project_life_cycles
+
+  belongs_to :color, optional: false
+
+  validates :name, presence: true
+  validates :type, inclusion: { in: %w[Stage Gate], message: :must_be_a_stage_or_gate }
+
+  def initialize(*args)
+    if instance_of? LifeCycle
+      # Do not allow directly instantiating this class
+      raise NotImplementedError, "Cannot instantiate the base LifeCycle class directly. " \
+                                 "Use Stage or Gate instead."
+    end
+
+    super
+  end
```

Comment:
I believe it is required, because otherwise the validations of the child classes can be circumvented. Let's consider the following scenario:
 1. Instantiate a `Project::LifeCycleStep.new`
 2. Assign attributes outside of the initialisation also including the `type: "Project::Stage"`
 3. The class of the already initialised object won't change to `Project::Stage` after the assignment, and the validations declared in the `Project::Stage` class won't be applied.
 4. An invalid object will be saved to the database.

```ruby
[1] pry(main)> lc = Project::LifeCycleStep.new
=> #<Project::LifeCycleStep:0x00000003144794d0
 id: nil,
 type: nil,
 start_date: nil,
 end_date: nil,
 active: false,
 position: nil,
 project_id: nil,
 definition_id: nil,
 created_at: nil,
 updated_at: nil>
[2] pry(main)> lc.assign_attributes(type: "Project::Stage", definition: Project::LifeCycleStepDefinition.first, project: Project.first)
  Project::LifeCycleStepDefinition Load (1.8ms)  SELECT "project_life_cycle_step_definitions".* FROM "project_life_cycle_step_definitions" ORDER BY "project_life_cycle_step_definitions"."id" ASC LIMIT $1  [["LIMIT", 1]]
  Project Load (0.5ms)  SELECT "projects".* FROM "projects" ORDER BY "projects"."id" ASC LIMIT $1  [["LIMIT", 1]]
=> nil
[3] pry(main)> lc.valid?
=> true
[4] pry(main)> lc
=> #<Project::LifeCycleStep:0x00000003144794d0
 id: nil,
 type: "Project::Stage",
 start_date: nil,
 end_date: nil,
 active: false,
 position: nil,
 project_id: 1,
 definition_id: 1,
 created_at: nil,
 updated_at: nil>
[5] pry(main)> lc.save
  TRANSACTION (11.8ms)  BEGIN
  Project::LifeCycleStep Load (18.0ms)  SELECT "project_life_cycle_steps".* FROM "project_life_cycle_steps" WHERE (1 = 1) AND ("project_life_cycle_steps"."position" IS NOT NULL) ORDER BY "project_life_cycle_steps"."position" DESC LIMIT $1  [["LIMIT", 1]]
  Project::LifeCycleStep Create (3.9ms)  INSERT INTO "project_life_cycle_steps" ("type", "position", "project_id", "definition_id", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"  [["type", "Project::Stage"], ["position", 1], ["project_id", 1], ["definition_id", 1], ["created_at", "2024-11-22 11:21:33.569747"], ["updated_at", "2024-11-22 11:21:33.569747"]]
  TRANSACTION (2.7ms)  COMMIT
=> true
[6] pry(main)> Project::LifeCycleStep.find(lc.id).valid? 
  Project::LifeCycleStep Load (12.4ms)  SELECT "project_life_cycle_steps".* FROM "project_life_cycle_steps" WHERE "project_life_cycle_steps"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
  Project Load (0.4ms)  SELECT "projects".* FROM "projects" WHERE "projects"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
  Project::LifeCycleStepDefinition Load (0.2ms)  SELECT "project_life_cycle_step_definitions".* FROM "project_life_cycle_step_definitions" WHERE "project_life_cycle_step_definitions"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
=> false

```

Reply:
It is still possible to create an invalid Stage if Gate is initialised and then switched dynamically - create a gate with only `start_date`/`date` and change type to Stage. So it is a problem of STI, if you really want to ensure no invalid instances, all validation should move from sub classes to base class and apply depending on the value of `type` attribute
```rb
gate = Project::Gate.new
gate.attributes = {start_date: Date.today, type: 'Project::Stage', definition: Project::LifeCycleStepDefinition.first, project: Project.first}
gate.valid? # => true
gate.save
Project::LifeCycleStep.find(gate.id).valid? # => false
Project::LifeCycleStep.find(gate.id).tap(&:validate).errors # => #<ActiveModel::Errors [#<ActiveModel::Error attribute=end_date, type=blank, options={}>]>
```

Reply:
I don't like the idea of moving class related validations into the parent model, it defeats the purpose of the whole structure.
I would rather add a validation the following validation to ensure no invalid objects can be created:

```ruby
class Project::Stage < Project::LifeCycleStep
  validates :type, inclusion: { in: %w[Project::Stage], message: :must_be_a_stage }
end

class Project::Gate < Project::LifeCycleStep
  validates :type, inclusion: { in: %w[Project::Gate], message: :must_be_a_gate }
end
```